### ISSUE TITLE
PHP 8.4 - Implicit nullability deprecation

### ISSUE THREAD

**Issue by Ayesh** [thecodingmachine/safe#442]
PHP 8.4 recently merged changes to emit a deprecation notice on function declared with an implicit nullable type.

All upstream functions in php-src already have the nullable types (`?string $var = null` and not `string $var = null`), so we will need to update our generator to account for this.

I submitted PR #441 to preview the expected changes.

**Comment by sebastianbergmann** [comment]
This library is used by @theofidry's `php-scoper` tool which is used in the release process of PHPUnit.

When I use `php-scoper` with PHP 8.4 then I get
```
PHP Fatal error:  Uncaught ErrorException: _HumbugBoxf238c5c0abfa\Safe\gmdate(): Implicitly marking parameter $timestamp as nullable is deprecated, the explicit nullable type must be used instead in phar:///usr/local/src/phpunit/tools/php-scoper/vendor/thecodingmachine/safe/deprecated/datetime.php:6
```

There must be an error handler, either in this library or in `php-scoper`, that escalates the deprecation to a fatal error.

**TL;DR:** As of right, I cannot release new versions of PHPUnit using PHP 8.4 due to this library not being compatible with PHP 8.4.

**Comment by staabm** [comment]
@sebastianbergmann @theofidry looking at this repo I can't see a error handler. 

since PHPUnit got new releases I guess this problem got somehow fixed for you?

**Comment by staabm** [comment]
the OPs problem got fixed by
- https://github.com/thecodingmachine/safe/commit/f87af02c112e4afdc42ee002e3ce86531df70ab5
- https://github.com/thecodingmachine/safe/commit/8a6679c47aaee74d3d90fe27df0c544335db019e

**Comment by sebastianbergmann** [comment]
> since PHPUnit got new releases I guess this problem got somehow fixed for you?

I downgraded PHP to roll the release of PHPUnit 11.4.4.

**Comment by jhdxr** [comment]
> the OPs problem got fixed by
> 
>     * [f87af02](https://github.com/thecodingmachine/safe/commit/f87af02c112e4afdc42ee002e3ce86531df70ab5)
> 
>     * [8a6679c](https://github.com/thecodingmachine/safe/commit/8a6679c47aaee74d3d90fe27df0c544335db019e)

Hi @staabm is there any chance that you have a plan to publish a release to include these changes? Last release is over one year ago.

**Comment by staabm** [comment]
We discuss it in https://github.com/thecodingmachine/safe/issues/499
